src/main/java/gregtech/common/items/behaviors/spray/AbstractSprayBehavior.java

```diff
@@ -0,0 +1,187 @@
+                    1.0f, 1.0f);
+        }
+        return result;
+    }
+
+    protected @NotNull EnumActionResult spray(@NotNull EntityPlayer player, @NotNull World world, @NotNull BlockPos pos,
+                                              @NotNull EnumFacing facing, @NotNull ItemStack sprayCan) {
+        if (!canSpray(sprayCan)) {
+            return EnumActionResult.PASS;
+        } else if (!player.canPlayerEdit(pos, facing, sprayCan)) {
+            return EnumActionResult.FAIL;
+        }
+
+        if (player.isSneaking()) {
+            if (world.getBlockState(pos).getBlock() instanceof BlockPipe<?, ?, ?>blockPipe) {
+                RayTraceResult hitResult = blockPipe.getServerCollisionRayTrace(player, pos, world);
+                if (hitResult != null) {
+                    EnumFacing hitSide = CoverRayTracer.determineGridSideHit(hitResult);
+                    IPipeTile<?, ?> firstPipe = blockPipe.getPipeTileEntity(world, pos);
+                    int color = getColorInt(sprayCan);
+                    if (hitSide != null && firstPipe != null && firstPipe.isConnected(hitSide) &&
+                            (firstPipe.isPainted() ? firstPipe.getPaintingColor() != color : color != -1)) {
+                        traversePipes(firstPipe, hitSide, player, sprayCan, color);
+                        return EnumActionResult.SUCCESS;
+                    }
+                }
+            }
+        }
+
+        ColoredBlockContainer colorContainer = ColoredBlockContainer.getContainer(world, pos, facing, player);
+        if (colorContainer.supportsARGB() ? colorContainer.setColor(world, pos, facing, player, getColorInt(sprayCan)) :
+                colorContainer.setColor(world, pos, facing, player, getColor(sprayCan))) {
+            onSpray(player, sprayCan);
+            return EnumActionResult.SUCCESS;
+        }
+
+        return EnumActionResult.PASS;
+    }
+
+    protected void traversePipes(@NotNull IPipeTile<?, ?> pipeTile, @NotNull EnumFacing facing,
+                                 @NotNull EntityPlayer player, @NotNull ItemStack sprayCan, int color) {
+        if (canPipeBePainted(pipeTile, color) && pipeTile.getNeighbor(facing) instanceof IPipeTile<?, ?>nextPipe) {
+            pipeTile.setPaintingColor(color);
+            onSpray(player, sprayCan);
+            pipeTile = nextPipe;
+        } else {
+            return;
+        }
+
+        for (int count = 1; count < getMaximumSprayLength(sprayCan) && canSpray(sprayCan); count++) {
+            if (canPipeBePainted(pipeTile, color)) {
+                pipeTile.setPaintingColor(color);
+                onSpray(player, sprayCan);
+            } else {
+                break;
+            }
+
+            if (pipeTile.getNumConnections() == 2) {
+                int connections = pipeTile.getConnections();
+                connections &= ~(1 << facing.getOpposite().getIndex());
+                for (EnumFacing other : EnumFacing.VALUES) {
```

Comment:
Didn't the superseded PR create a custom pipe net walker to walk the pipe network, with a limit on the number of steps based on max uses left of the can. Why did we switch to just iterating over every direction? Also, I think this was causing the case where a `+` would not get fully colored.